shared/utils/nric-validation.ts

```diff
@@ -1,41 +1,33 @@
+type NricParts = {
+  prefix: string
+  digits: string
+  checksum: string
+}
+
+const NRIC_FORMAT = /^(?<prefix>[STFG])(?<digits>\d{7})(?<checksum>[A-Z])$/
+
 /**
  * Validates whether a provided string value adheres to the UIN/FIN format
  * as provided on the Singapore Government's National Registration Identity Card.
  * @param value The value to be validated
  */
 export const isNricValid = (value: string): boolean => {
-  return isFormatValid(value) && isChecksumValid(value)
-}
+  const parsed = value?.toUpperCase().match(NRIC_FORMAT)
 
-/**
- * Tests whether a provided string value obeys a simple format check
- * @param value The value to be validated
- */
-const isFormatValid = (value: string): boolean => {
-  return /^([STFGstfg]{1})([0-9]{7})([A-Za-z]{1})$/.test(value)
-}
+  if (!parsed) return false
+
+  const { prefix, digits, checksum } = parsed.groups as NricParts
 
-/**
- * Algorithm to test whether the NRIC checksum is valid
- * @param value The value to be validated
- */
-const isChecksumValid = (value: string): boolean => {
   // http://www.ngiam.net/NRIC/NRIC_numbers.pdf
-  value = value.toUpperCase()
-  const prefix = value.charAt(0)
-  const suffix = value.charAt(value.length - 1)
   const coefficients = [2, 7, 6, 5, 4, 3, 2]
-  const constant = prefix === 'S' || prefix === 'F' ? 0 : 4
-  const coding =
-    prefix === 'S' || prefix === 'T'
-      ? ['J', 'Z', 'I', 'H', 'G', 'F', 'E', 'D', 'C', 'B', 'A']
-      : ['X', 'W', 'U', 'T', 'R', 'Q', 'P', 'N', 'M', 'L', 'K']
-  const sum = value
-    .substring(1, value.length - 1)
-    .split('')
-    .reduce(function (sum, digit, idx) {
-      sum += parseInt(digit) * coefficients[idx]
-      return sum
-    }, constant)
-  return suffix === coding[sum % 11]
+  const startConstant = prefix === 'S' || prefix === 'F' ? 0 : 4
+  const checksumEncoding =
+    prefix === 'S' || prefix === 'T' ? 'JZIHGFEDCBA' : 'XWUTRQPNMLK'
+
+  const sum = coefficients.reduce(
```

Comment:
Could I suggest adding a comment so its clear to future devs what this line does?

```suggestion
  // Multiply coefficients and digits element-wise and then sum the result 
  const sum = coefficients.reduce(
```

Reply:
In this case, I'm thinking the code is very short using simple well-understood operations, so I reckon it's rather easy to understand on its own, and a comment that duplicates what the code does is probably unnecessary.

That said, the code itself maybe could be made a little bit more self-explanatory 🤔 . Renaming `coefficients` into `weights`, and `sum` into `weightedSum` might make clearer that it's just that: simple code to compute a weighted sum 😅 . I'll add a commit to do that!

Reply:
Added commit to rename variables! I think the code reads pretty clearly now 👍 😃